### PR TITLE
Permissioned Slot and Model.Access

### PR DESCRIPTION
**Pull request by chreekat**
Applies #642's treatment of getVolume to getSlot. Then it takes it one step further, combining the implementations and giving the concept a new name: Model.Access.

### CODE REVIEW

File: src/Controller/Slot.hs [hunk 1]
```diff
@@ -41,22 +41,43 @@ import Model.Volume
...
+    -- ^ Requested permission
+    -> Id Slot
+    -- ^ Slot to look up
+    -> Handler Slot
+    -- ^ The slot, as requested (or a short-circuited error response)
+getVolumeSlot volId requestedPerm sId = do
+    s <- getSlot requestedPerm sId
+    if volumeId (volumeRow (containerVolume (slotContainer s))) == volId
+        then pure s
+        else result =<< peeks notFoundResponse
```
Comment (kanishka-azimi) [inline src/Controller/Slot.hs]: eventually, a type that encapsulates this like (Id Slot, Id Volume)

File: test/ModelTest.hs [hunk 2]
```diff
@@ -282,30 +282,39 @@ test_11 = Test.stepsWithTransaction "test_11" $ \step cn2 -> do
...
+        )
+        aiCtxt
     step "Then the public can't view the container"
-    -- Implementation of getSlot PUBLIC
-    mSlotForAnon <- runWithNoIdent cn2 (lookupContainerSlot cid)
-    isNothing mSlotForAnon @? "expected slot lookup to find nothing"
+    mSlotForAnon <- runWithNoIdent
+        cn2
+        (requestSlot PermissionPUBLIC (containerSlotId cid))
+    mSlotForAnon @?= LookupFailed
```
Comment (kanishka-azimi) [inline test/ModelTest.hs]: runWithNoIdent - move towards the pattern you identifid
(flip runReaderT) makeNoIdentCtxt (do
   ...
)

File: test/ModelTest.hs [hunk 3]
```diff
@@ -282,30 +282,39 @@ test_11 = Test.stepsWithTransaction "test_11" $ \step cn2 -> do
...
-    -- Implementation of getSlot PUBLIC
-    Just slotForAnon <- runWithNoIdent cn2 (lookupContainerSlot cid)
+    slotForAnon <- runWithNoIdent
+        cn2
+        (requestSlot PermissionPUBLIC (containerSlotId cid))
     step "Then the public can't see protected parts like the detailed test date"
-    (encode . getContainerDate . slotContainer) slotForAnon @?= "2017"
+    case slotForAnon of
+        LookupFailed -> assertFailure "Lookup failed"
+        RequestDenied -> assertFailure "Access denied"
```
Comment (kanishka-azimi) [inline test/ModelTest.hs]: usually expected ... , actual .... in message

File: src/Controller/Slot.hs (hunk 1 above)
Comment (kanishka-azimi) [reply src/Controller/Slot.hs]: ```data SlotLookupKey = DirectSlotId ... | VolumeSlotId ... ```

File: src/Model/Slot.hs [hunk 4]
```diff
@@ -17,7 +18,38 @@ import Model.Identity
...
+requestSlot
+    :: (MonadDB c m, MonadHasIdentity c m)
+    => Permission
+    -> Id Slot
+    -> m (RequestResult Slot)
+requestSlot requestedPerm = fmap (maybe LookupFailed mkRequest) . lookupSlotP
+  where
+    mkRequest :: Permissioned Slot -> RequestResult Slot
+    mkRequest =
+        maybe RequestDenied RequestResult . requestAccess requestedPerm
```
Comment (kanishka-azimi) [inline src/Model/Slot.hs]: nested maybe feels a little spread out. I'd probably have one case or one nested case

File: src/Model/Slot.hs [hunk 5]
```diff
@@ -17,7 +18,38 @@ import Model.Identity
...
+-- | Lookup a Slot by its Id, requesting the given permission.
+--
+-- NOTE: Intentionally implemented exactly like requestVolume. Implementations
+-- should be collected in a single module and merged.
+requestSlot
+    :: (MonadDB c m, MonadHasIdentity c m)
+    => Permission
+    -> Id Slot
+    -> m (RequestResult Slot)
+requestSlot requestedPerm = fmap (maybe LookupFailed mkRequest) . lookupSlotP
```
Comment (kanishka-azimi) [inline src/Model/Slot.hs]: I'd name the id slot parameter

File: src/Model/Slot.hs [hunk 6]
```diff
@@ -17,7 +18,38 @@ import Model.Identity
...
+  where
+    mkRequest :: Permissioned Slot -> RequestResult Slot
+    mkRequest =
+        maybe RequestDenied RequestResult . requestAccess requestedPerm
+    --
+    lookupSlotP
+        :: (MonadDB c m, MonadHasIdentity c m)
+        => Id Slot
+        -> m (Maybe (Permissioned Slot))
+    lookupSlotP = fmap (fmap wrapPermission) . lookupSlot
```
Comment (kanishka-azimi) [inline src/Model/Slot.hs]: same feeling as before (:

File: test/ModelTest.hs (hunk 3 above)
Comment (chreekat) [reply test/ModelTest.hs]: Yeah that would be ideal. Can do.

File: src/Model/Slot.hs (hunk 6 above)
Comment (chreekat) [reply src/Model/Slot.hs]: There's only one of them now. :)

**Comment by chreekat** [comment]
> Overall, this PR and the last few PRs around controller.permission, getVolume, and getSlot, I would have aimed for less abstractions. [...] There is definitely a need for nicer abstractions. I just don't think you have enough context yet to know what those are.

You are correct that I don't have the whole picture/full context. If you think I'm going too far off the rails, I'm happy to revert changes.

When I am adding new features or fixing bugs, I try my absolute best to minimize diffs and introduce no new abstractions. In this case, though, my whole purpose is to refactor code under test with no change of behavior. I'm taking more liberties.

I'm trying to avoid perfectionism, too. In general, I don't think it's possible to keep the full picture of a software project in my head, and I don't want to rely on that hypothetical capability. The best I can do is create good abstractions for the specific piece I'm looking at. If I find an impedance mismatch upon moving to the next task, I can refactor again.  It is fun and easy to refactor with Haskell, and I want to take advantage of that.

Without a doubt, though, you have a better picture of things than I do! You've been dropping hints about the next pieces to look at (* versus *Row, *JSONField functions)... write more toehold tests, and I'll start ripping up those, too. ;)

Of course, one very real danger is creating a confusion of abstractions. Now that I've done getVolume and getSlot in a particular way, a newcomer might notice getContainer is implemented differently and not know why. I'd still rather move forward with piecewise refactorings, however. Comments and issues can keep track of works in progress, and even provide entry points for volunteers.

**Comment by kanishka-azimi** [comment]
Don't revert anything. The new code is better, and feel free to keep applying this approach to similar functions.

I wrote my note for consideration when you start a new pass of refactoring (e.g. functions that create a volume, create a slot, etc).

> When I am adding new features or fixing bugs, I try my absolute best to minimize diffs and introduce no new abstractions. In this case, though, my whole purpose is to refactor code under test with no change of behavior.

It's funny. I work in the exact opposite style (:. When I am on a feature or bug, I feel like it focuses me on my future trajectory I want and I usually deal with the change at hand first, but quickly follow-up with huge refactorings. I guess we each have our own styles.

**Pull request merged**